### ISSUE TITLE
Selectors with no dependencies have `never` input parameter

### ISSUE THREAD

**Issue by chbdetta** [reduxjs/reselect#538]
Getting this with the latest 4.1.1 version

```TypeScript
import {createSelector} from 'reselect';

const mySelector = createSelector([], () => ({
    foo: 'foo'
}))

mySelector({});
// Error, {} is not assignable to never
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwMZQKYEMaoMqoDapIzQC+cAZlBCHAORoDO+hMtA3ALABQ3SEAdg3ggAnrgJFocALxwUGLOJbQAFAG0AugBo4KgJQyAfLoTc45ihAgAuOuSu1uJPXu7dRSyVBUJnnLgD0AXAAegD83EFwQpiotvyoAG6oUEA

**Comment by klimashkin** [comment]
Seems too restrictive. Should it be `unknown` instead?

**Comment by markerikson** [comment]
Legit question: why would you _ever_ have a selector here with no dependencies? The whole point of `createSelector` is to memoize on the input values.

**Comment by klimashkin** [comment]
Good question, @markerikson!
We use it to cache the result of a function forever, but lazily only after it gets called for the first time. We could probably use another utils for that, but a selector with no dependencies already works very well for that case.

Let's say you have some kind of `intl` global  function in an object prop (which might be a config for some component), but the result of that function depends on an initial api call (like user settings). You could write it like
```js
import intl from 'intl';

const tableConfig = {
  name: intl('List'),
  ...hundred of other props...
}
```
So intl() in that case will be called immediately upon module execution, and we don't want that, because we still don't have critical information in place. We want that config to be created only when we are ready to render actual page, and that happens after we fetch user settings.

So, we simply wrap it into a selector which makes it a function, which creates the object once (and only when we are ready) and caches it forever.
```js
import intl from 'intl';

const tableConfig = createSelector([], () => ({
  name: intl('List'),
  ...hundred of other props...
}));
```
 
It's similar to having `useMemo` with `[]`, but outside of the React.

There are different ways to approach that problem, but `createSelector` was the easiest one to add, because we already use `createSelector` in the application.
And conceptually in JS there is no limitation of calling a function with a number of parameters bigger than then number of arguments, down to zero arguments. So a selector with zero dependencies should still be able to participate in dependencies of other selectors, which might or might not get state and props arguments, which I think makes `unknown` a better candidate here than `never` ("Be liberal in what you accept and strict in what you produce")

**Comment by markerikson** [comment]
I think what you might actually want in this case is not even `createSelector`, but the `defaultMemoize` function itself. Try passing your "run this once" function to `defaultMemoize` instead and see what happens?

**Comment by eXamadeus** [comment]
Technically, the type is correct...the selector is expecting `never` because there is no dependency from which it can gather type information. So it assumes, correctly, there is no type possible to pass into a non-existent dependency.

I don't think there is a fix for this since it's actually doing what you're asking it. I think this might just be a situation where casting `{} as never` will get you through.

You can do this:
```ts
const mySelector = createSelector([], () => ({
    foo: 'foo'
}))

const NEVER = null as never
mySelector(NEVER)
mySelector({} as never)
mySelector(null as never)
// etc..
```

or if you want to pass in `null` directly, this could work:

```ts
const myNullSelector = createSelector((state: null) => {}, () => ({
    foo: 'foo'
}))

myNullSelector(null)
```

The types, as they are, very specifically pull out the information handed into the `createSelector` function, so there's not exactly a typing problem here. In fact, attempting to fix this will result in breaking a _lot_ of other working type checks that happen. It's definitely not something we can easily support (if we can at all).

An example of what breaks, if we make the following change to the `Selector` type:

```ts
export type Selector<
  S = any,
  R = unknown,
  P extends never | readonly any[] = any[]
> = [P] extends [never]
  ? [S] extends [never]
    ? (state?: S) => R
    : (state: S) => R
  : (state: S, ...params: P) => R
```

Now if the `S` type (meaning state type) is inferred to `never` it will make the resultant signature `(state?: never) => R`, which will work for the case you mentioned above. Where it fails is here:

```ts
  const selector = createSelector(
    (state: string) => 1,
    (state: number) => 2,
    (...args) => 0
  )

  // @ts-expect-error
  selector('foo')
  // @ts-expect-error
  selector() // this will erroneously NOT show a type error now...which isn't good.
  // @ts-expect-error
  selector(5)
```

### TLDR

This is not something we can support currently, nor am I convinced we should. I recommend `{} as never` since it's already a kinda weird thing anyway. Or better yet, what @markerikson suggested:

```ts
import { defaultMemoize } from "reselect"

const memoizedFunc = defaultMemoize(() => ({
  foo: "foo",
}))

memoizedFunc()
```

**Comment by markerikson** [comment]
Based on that answer, I think we can go ahead and close this. Let us know if there are further issues!

**Comment by klimashkin** [comment]
@markerikson @eXamadeus I guess it was fixed in 4.1.3, now empty dependencies work and the type is `() => unknown`

**Comment by markerikson** [comment]
Huh, I guess some of the parameter merging fixes I did changed behavior here.

Can you also double-check the https://github.com/reduxjs/reselect/releases/tag/v4.1.4 release I did just now?

**Comment by klimashkin** [comment]
Just checked 4.1.4, in this case behavior is similar to 4.1.3!
